_posts/2022/2022-10-31-use-zero-for-integer-float-comparison.md

```diff
@@ -0,0 +1,119 @@
+
+An example of this is the common programming comparison of numeric variable to zero.
+
+See the documentation of the `.zero?` method on [`Integer`](https://ruby-doc.org/core-3.1.2/Integer.html#zero-3F-method), [`Float`](https://ruby-doc.org/core-3.1.2/Float.html#zero-3F-method), [`Numeric`](https://ruby-doc.org/core-3.1.2/Numeric.html#zero-3F-method) and [`BigDecimal`](https://ruby-doc.org/stdlib-3.1.2/libdoc/bigdecimal/rdoc/BigDecimal.html#zero-3F-method).
+
+
+## Instead of...
+
+...checking whether a value is zero in a conditional, how other languages might do it:
+
+```ruby
+if number == 0
+  :yes
+else
+  :no
+end
+```
+
+
+## Use...
+
+...the built-in `#zero?` method on all numeric types:
+
+```ruby
+if number.zero?
+  :yes
+else
+  :no
+end
+```
+
+
+## Why?
+
+The Ruby-ish syntax in this case is clearer and less error prone.
+
+It is very easy to type `number = 0` and thus assign a variable in the conditional rather than compare (although you might use tests to protect yourself from this particular error).
+
+
+## Why not?
+
+This is a habit I still struggle with after 15 years of Ruby programming. The `== 0` syntax is how _many_ other similar languages do comparison and is deeply ingrained in my fingers!
+
+You might get folks pushing back for “performance” reasons.
+
+```ruby
+require "benchmark/ips"
+
+Benchmark.ips do |x|
+  x.report("1 == 0") { 1 == 0 } #=> false
+  x.report("0 == 0") { 0 == 0 } #=> true
+  x.report("1.zero?") { 1.zero? } #=> false
+  x.report("0.zero?") { 0.zero? } #=> true
+
+  x.report("1.0 == 0") { 1.0 == 0 } #=> false
+  x.report("0.0 == 0") { 0.0 == 0 } #=> true
+  x.report("1.0.zero?") { 1.0.zero? } #=> false
+  x.report("0.0.zero?") { 0.0.zero? } #=> true
+end
+```
+
+For `Integer`:
+
+<table>
+<tr>
+  <th>1 == 0</th>
+  <td class="text-right">29.049M <small>(± 0.7%)</small> i/s</td>
+</tr>
+<tr>
+  <th>0 == 0</th>
+  <td class="text-right">28.972M <small>(± 0.3%)</small> i/s</td>
+</tr>
+<tr>
+  <th>1.zero?</th>
+  <td class="text-right">23.009M <small>(± 1.0%)</small> i/s</td>
+</tr>
+<tr>
+  <th>0.zero?</th>
+  <td class="text-right">22.860M <small>(± 1.7%)</small> i/s</td>
+</tr>
+</table>
+
+For `Float`:
+
+<table>
+<tr>
+  <th>1.0 == 0</th>
+  <td class="text-right">18.768M <small>(± 2.0%)</small> i/s</td>
+</tr>
+<tr>
+  <th>0.0 == 0</th>
+  <td class="text-right">19.066M <small>(± 0.5%)</small> i/s</td>
+</tr>
+<tr>
+  <th>1.0.zero?</th>
+  <td class="text-right">22.784M <small>(± 0.6%)</small> i/s</td>
+</tr>
+<tr>
+  <th>0.0.zero?</th>
+  <td class="text-right">22.841M <small>(± 0.8%)</small> i/s</td>
+</tr>
+</table>
+
+The benchmark shows that for `Integer`s the `==` syntax is faster, but for `Float`s the `.zero?` syntax wins out.
+
+However the important thing to note is that in all cases (on my machine) you get about 20 million executions per second so in _your_ code consider the readability over the performance implications!
```

Comment:
I think this summary could be said more clearly and remove reference to your machine if possible to make the argument more strongly.

Reply:
```suggestion
However the important thing to note is that in all cases you get millions of executions per second so in _your_ code you should emphasise the readability over any perceived performance implications!
```